MyList/src/main/java/com/example/mylist/MyList.java

```diff
@@ -0,0 +1,451 @@
+            result[i] = elementData[i];
+
+        }
+        return result;
+    }
+
+    //. 1. 매개변수에 어떤타입이던 리턴시에 따라갈줄알았는데 타입캐스팅이 필요하다고?
+    /*
+    실행부...이해 100퍼 못했음
+    String[] arr = new String[2];
+    arr = (String[]) list.toArray(arr);
+     */
+
+    //TODO 제네릭이해도가 낮다 멘토님의 도움을 후에 요청해보자
+    @Override
+    public <T> T[] toArray(T[] a) {
+        //이건되는데
+        if(a.length >= size) {
+            System.arraycopy(elementData, 0, a, 0,size);
+            return a;
+        }
+        //이게 안되네
+        if(a.length < size) {
+            Object result = new Object[size];
+            System.arraycopy(elementData, 0, result, 0,size);
+            return (T[]) result;
+        }
+
+
+
+        return a;
+    }
+    // toArray 끝 ==========================================================================
+
+
+    // remove 시작 ==========================================================================
+    @Override
+    public boolean remove(Object o) {
+        for(int i=0; i<elementData.length; i++) {
+            if(elementData[i].equals(o)) {
+                System.arraycopy(elementData, i+1, elementData, i, size-1);
+                size--;
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    @Override
+    public boolean removeAll(Collection<?> c) {
+
+        boolean flag = false;
+        //반목이 많아지면 성능상 별로일거같은데 다른 방법을 생각해보았으나 결국 Stream이고
+        //Sream역시도 결국 내부 반복이고 반복문은 피할수없을것같다
+        // 반복문이 성능에 문제를 주긴하지만 내가 생각한것 만큼 많이 잡아먹지는 않는다고는한다
+        for(Object o : c) {
+            if(contains(o)) {
+                //length가 아니라 size로 해야한다
+                for(int i=0; i<size; i++) {
+                    if(elementData[i].equals(o)) {
+                        System.arraycopy(elementData, i+1, elementData, i, size-1);
+                        //이거때문에 remove()를 못쓰다니...
+                        elementData[size-1] = null;
+                        size--;
+                    }
+                }
+            }
+            flag = true;
+        }
+        return flag;
+    }
+
+    @Override
+    public E remove(int index) {
+        checkIndex(index);
+
+        E before = elementData(index);
+        System.arraycopy(elementData, index+1, elementData, index, size-1);
+        size--;
+
+        return before;
+    }
+
+    // remove 끝 ==========================================================================
+
+    @Override
+    public boolean containsAll(Collection<?> c) {
+
+        for(Object o : c) {
+            if(!contains(o)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public boolean retainAll(Collection<?> c) {
+
+        //이터레이터를 사용한다?
+        //우선 시간상 보류
+        return false;
+    }
+
+    @Override
+    public void clear() {
+        //요소를 제거하는거니깐 null로 초기화 리스트의 크기는 유지되겠지만
+        // 참조요소를 없애면 GC가 처리해줄라나?
+        for(int i=0; i<size; i++) {
+            elementData[i] = null;
+        }
```

Comment:
아예 배열을 새로 생성하여 할당하는 방법이 있죠.
위 코드방식은 O(n)이겠지만, 신규 배열을 생성한다면 O(1)입니다.

```suggestion
        
            elementData = new Object[size];
        
```